src/core/vector/qgsvectorlayer.cpp

```diff
@@ -1032,22 +1032,34 @@ QgsRectangle QgsVectorLayer::extent() const
   if ( !isSpatial() )
     return rect;
 
-  if ( !mValidExtent2D && mLazyExtent2D && mReadExtentFromXml && !mXmlExtent2D.isNull() )
+  if ( mDataProvider && mDataProvider->isValid() && mDataProvider->flags() & Qgis::DataProviderFlag::FastExtent2D )
```

Comment:
I'm always nervous about operator priority ...
```suggestion
  if ( mDataProvider && mDataProvider->isValid() && (mDataProvider->flags() & Qgis::DataProviderFlag::FastExtent2D) )
```